src/languages/golfscript/emit.ts

```diff
@@ -1,133 +1,146 @@
+    return ["{", children.map((stmt) => emitStatement(stmt, baseExpr)), "}"];
   }
 
-  return ["{", children.map((stmt) => emitStatement(stmt, baseExpr)), "}"];
-}
-
-function emitStatement(stmt: IR.Expr, parent: IR.Node): TokenTree {
-  switch (stmt.kind) {
-    case "Block":
-      return emitMultiExpr(stmt, parent);
-    case "ImportStatement":
-      return [stmt.name, ...stmt.modules]; // TODO the ... could be avoided if TokenTree was made readonly??
-    case "WhileLoop":
-      return [
-        emitMultiExpr(stmt.condition, stmt),
-        emitMultiExpr(stmt.body, stmt),
-        "while",
-      ];
-    case "ForRange":
-      return [
-        emitExpr(stmt.high),
-        ",",
-        isIntLiteral(stmt.low, 0n) ? [] : [emitExpr(stmt.low), ">"],
-        isIntLiteral(stmt.increment, 1n) ? [] : [emitExpr(stmt.increment), "%"],
-        "{",
-        ":",
-        emitExpr(stmt.variable),
-        ";",
-        emitMultiExpr(stmt.body, stmt),
-        "}",
-        "%",
-      ];
-    case "ForEach":
-      return [
-        emitExpr(stmt.collection),
-        "{",
-        ":",
-        emitExpr(stmt.variable),
-        ";",
-        emitMultiExpr(stmt.body, stmt),
-        "}",
-        "%",
-      ];
-    case "IfStatement":
-      return [
-        emitExpr(stmt.condition),
-        emitMultiExpr(stmt.consequent, stmt),
-        stmt.alternate !== undefined
-          ? emitMultiExpr(stmt.alternate, stmt)
-          : "{}",
-        "if",
-      ];
-    case "Variants":
-    case "ForEachKey":
-    case "ForEachPair":
-    case "ForCLike":
-      throw new EmitError(stmt);
-    default:
-      return emitExpr(stmt);
+  function emitStatement(stmt: IR.Expr, parent: IR.Node): TokenTree {
+    switch (stmt.kind) {
+      case "Block":
+        return emitMultiExpr(stmt, parent);
+      case "ImportStatement":
+        return [stmt.name, ...stmt.modules]; // TODO the ... could be avoided if TokenTree was made readonly??
+      case "WhileLoop":
+        return [
+          emitMultiExpr(stmt.condition, stmt),
+          emitMultiExpr(stmt.body, stmt),
+          "while",
+        ];
+      case "ForRange": {
+        if (stmt.inclusive) throw new EmitError(stmt, "inclusive");
+        const isNaturalType = (x: Expr) =>
+          isSubtype(getType(x, program), integerType(0, "oo"));
+        if (
+          isNaturalType(stmt.low) &&
+          isNaturalType(stmt.high) &&
+          isNaturalType(stmt.increment)
+        )
+          return [
+            emitExpr(stmt.high),
+            ",",
+            isIntLiteral(stmt.low, 0n) ? [] : [emitExpr(stmt.low), ">"],
+            isIntLiteral(stmt.increment, 1n)
+              ? []
+              : [emitExpr(stmt.increment), "%"],
+            "{",
+            ":",
+            emitExpr(stmt.variable),
+            ";",
+            emitMultiExpr(stmt.body, stmt),
+            "}",
+            "%",
+          ];
+        throw new EmitError(stmt, "potentially negative");
```

Comment:
```suggestion
        return [
          isNaturalType(stmt.low)
            ? [
                emitExpr(stmt.high),
                ",",
                isIntLiteral(stmt.low, 0n) ? [] : [emitExpr(stmt.low), ">"],
              ]
            : [
                emitExpr(stmt.high),
                emitExpr(stmt.low),
                "-",
                ",",
                "{",
                emitExpr(stmt.low),
                "+",
                "}",
                "%",
              ],
          isIntLiteral(stmt.increment, 1n)
            ? []
            : [emitExpr(stmt.increment), "%"],
          "{",
          ":",
          emitExpr(stmt.variable),
          ";",
          emitMultiExpr(stmt.body, stmt),
          "}",
          "%",
        ];
```
This handles a negative low and/or high bound. Negative increments function as intended, or at least in the same way as Python. If this is undesired behavior it should probably be impossible to use a negative step regardless of lang. This section of new code:
```
emitExpr(stmt.high),
emitExpr(stmt.low),
"-",
```
should emit a `sub` polyGolfOp acting on high and low, or whatever the best way is to enable the subtraction to be simplified, I'm not sure how to go about that though.